src/app/zones/data/index.ts

```diff
@@ -1,36 +1,87 @@
-import { getIsConnected } from '@/app/subscriptions/data'
-import type { ZoneOverview } from '@/types/index.d'
+import type { PaginatedApiListResponse as CollectionResponse } from '@/types/api.d'
+import type {
+  ZoneOverview as PartialZoneOverview,
+  ZoneInsight as PartialZoneInsight,
+  Zone as PartialZone,
+  KDSSubscription,
+} from '@/types/index.d'
 import { get } from '@/utilities/get'
 
-// TODO(jc): These two are extremely similar, see if we can merge
-export function getZoneControlPlaneEnvironment(zoneOverview: ZoneOverview): string {
-  // TODO(jc): is it ok to use the config from the first subscription we find here?
-  for (const subscription of zoneOverview.zoneInsight?.subscriptions ?? []) {
-    if (subscription.config) {
-      return JSON.parse(subscription.config).environment
-    }
-  }
-  return ''
+export type Zone = PartialZone & {
+  enabled: boolean
+}
+export type ZoneInsight = PartialZoneInsight & {
+  connectedSubscription?: KDSSubscription
+  config: Record<string, unknown>
+  authenticationType: string
+  environment: string
+}
+export type ZoneOverview = PartialZoneOverview & {
+  zoneInsight?: ZoneInsight
+  zone: Zone
+  state: 'online' | 'offline' | 'disabled'
 }
-export function getZoneDpServerAuthType(zone: ZoneOverview): string {
-  const subscriptions = zone.zoneInsight?.subscriptions ?? []
-  if (subscriptions.length > 0) {
-    const lastSubscription = subscriptions[subscriptions.length - 1]
-    if (lastSubscription.config) {
-      const parsedConfig = JSON.parse(lastSubscription.config)
-      return get(parsedConfig, 'dpServer.auth.type', '')
+
+export const Zone = {
+  fromObject: (item: PartialZone): Zone => {
+    return {
+      ...item,
+      enabled: !(item.enabled === false),
     }
-  }
-  return ''
+  },
 }
-// TODO(jc): end
+export const ZoneInsight = {
+  fromObject: (item?: PartialZoneInsight): ZoneInsight | undefined => {
+    // if item isn't set don't even try augmenting things
+    return isSet<PartialZoneInsight>(item)
+      ? ((item) => {
+        // figure out the connectedSubscription by looking at the connectTime
+        // and disconnectTime of the last subscription
+        const connectedSubscription: KDSSubscription | undefined = !Array.isArray(item.subscriptions)
+          ? undefined
+          : [item.subscriptions[item.subscriptions.length - 1]].find((item) => item.connectTime?.length && !item.disconnectTime)
```

Comment:
**Issue**: We should use the last subscription. It’s the most recent one. We used to try and find the first one that has no `disconnectTime`, but that’s not necessary. 

And let’s avoid this "array with one entry plus find/every/some method" pattern just to get to a one-line expression. The following is much, much more clear:

```suggestion
        const subscriptions = item.subscriptions ?? []
        const connectedSubscription: KDSSubscription | undefined = subscriptions[subscriptions.length - 1]
```

Reply:
Are you absolutely sure the last subscription always has a `connectTime` and never has a `disconnectTime`? At the very least the types say that `connectTime` can be `undefined`? In which case I would guess it hasn't connected?

Reply:
From a conversation offline:

1. Whilst connectTime in our frontend has always been marked as potentially being unset. `connectTime` is guaranteed to be set, so we should amend that throughout the application (and therefore after this PR)
2. It is not guaranteed that the last subscription doesn't have a `disconnectTime`, all subscriptions can have a disconnectTime when the CP disconnects all the subscriptions (from what I understand from the offline convo with backend folks)

Let me know if I understood this wrong ^

Reply:
Also to be clear, I'm keeping the code using the same style that it was using before. I personally think this is clearer than your suggestion and using less `?-y` things. I may change this in pt2 of this (a follow up PR though), but I would probably still use `.find` seeing as it returns `undefined` and therefore `connectedSubscription` is less likely to be set to `null`.

With this comment and my previous comment, I'd like to resolve this, let me know.

Reply:
> 2\. It is not guaranteed that the last subscription doesn't have a `disconnectTime`

Just to be explicit: this is normal and should be expected. The last subscription _may_ have a disconnect time; all prior subscriptions _must_ have a disconnect time.

---

Feel free to resolve this. I really think this style of expression is problematic and unnecessarily complex, and it should be changed. I hope you do this in that follow up. At the very least, please consider the following suggestion: should the subscriptions ever come in as an empty array, this code will cause an exception and using `Array.prototype.at` exposes the issue to the type system (see explanation below).

```suggestion
        const connectedSubscription: KDSSubscription | undefined = !Array.isArray(item.subscriptions)
          ? undefined
          : [item.subscriptions.at(item.subscriptions.length - 1)].find((item) => item.connectTime?.length && !item.disconnectTime)
```

---

For the sake of explaining why I think it’s problematic:

> Also to be clear, I'm keeping the code using the same style that it was using before. I personally think this is clearer than your suggestion and using less `?-y` things. I may change this in pt2 of this (a follow up PR though), but I would probably still use `.find` seeing as it returns `undefined` and therefore `connectedSubscription` is less likely to be set to `null`.

I do really feel very strongly about this and want it to be changed here.

The code in the PR currently uses this expression:

```ts
const connectedSubscription: KDSSubscription | undefined = !Array.isArray(item.subscriptions)
  ? undefined
  : [item.subscriptions[item.subscriptions.length - 1]].find((item) => item.connectTime?.length && !item.disconnectTime)
```

This expression is semantically rather complex and hides a possible bug:

First: `item.subscriptions[item.subscriptions.length - 1]` appears to be of type `KDSSubscription` but should more accurately appear to be of type `KDSSubscription | undefined`. That’s because we don’t have `noUncheckedIndexedAccess` enabled. Replacing it with `item.subscriptions.at(item.subscriptions.length - 1)` illustrates the problem with that: suddenly `item` in the `find` call is `KDSSubscription | undefined` and `item.connectTime` becomes an error.

This code would lead to an exception if `item.subscriptions` would be an empty array. Currently that wouldn’t happen, but that might change very easily.

Second: With `[item.subscriptions.at(item.subscriptions.length - 1)].find((item) => item.connectTime?.length && !item.disconnectTime)` we’re creating an array of finite length 1 to then call `Array.prototype.find` just so we can leverage the fact that `find` returns `undefined` when it doesn’t find a match. That’s unnecessary.

We could make all of that extremely transparent with the change I suggested:

```suggestion
        const lastSubscription = (item.subscriptions ?? []).at(-1)
        const connectedSubscription = lastSubscription && !lastSubscription.disconnectTime ? lastSubscription : undefined
```

It also maintains that the "empty value" is `undefined` and performs the check for the connected subscription in a direct manner without utilizing `find` on a one-element array.

> using less `?-y` things

It uses the exact same amount of ternary expressions (1). Also, ternary expressions are great because they make it more obvious what the returned value can be. The one-element array and find style makes that needlessly obscure.

Reply:
I just traced back a few other things mentioned here

> This expression is semantically rather complex and hides a possible bug:
> This code would lead to an exception if item.subscriptions would be an empty array. Currently that wouldn’t happen, but that might change very easily.

https://github.com/kumahq/kuma-gui/blob/e490ea4c0f84a30e040fa2008e6283caabc811c0/src/types/index.d.ts#L46

`subscriptions` here is always set according to our types?

The only way this could have been `undefined` previously to this PR is because `zoneIngress` could be `undefined`, but in this PR we check for that before even getting to this code, so that case can't happen here. i.e. As things stand right now, `subscriptions` cannot be `undefined`.

If anything we need to remove the `Array.isArray` check, but ideally I wanted to make sure the rest of the areas we need this sort of thing was looked at before a final nip and tuck.

As mentioned above I have a PR for zone-ingress coming up, which currently has subscriptions as a child of an `any` definition, this is using `Array.slice` and `find` to always use an array and return an `item | undefined`.

---

🤔  I'm going to move this approach over here now rather than later as I'm "pretty sure" that it will end up being like this throughout the application, plus bringing this over here now instead of later I think resolves your issue. Please note tho this amend will continue to have a `Array.isArray` even though `subscriptions` here is always supposed to be an array, and it 'may' change again later depending on what the rest ends up looking like.

Resolving this one, but let me know if I need to reopen.